// @wix/wix-data-platformized-provider/dist/esm/platformized-data-response.js
import {
    __spreadArray
} from "tslib";
import {
    messages,
    wixDataError,
    codes
} from '@wix/wix-data-errors';
import {
    parseDates
} from '@wix/wix-data-utils';
import mapErrorCode from './platformized-code-mapping';
var HooksErrorCode = 'WDE0078';
var PayloadTooLargeCode = 'WDE0109';
var ValidationErrorCode = 'WDE0080';
export var handleJsonResponse = function(trace) {
    return function(response) {
        var _a;
        return trace('platformized-data-response', {
            requestId: (_a = response.headers) === null || _a === void 0 ? void 0 : _a['x-wix-request-id'],
        })(function() {
            return parseDates(response.data);
        });
    };
};
export var handleJsonErrorResponse = function(trace) {
    return function(error) {
        var _a, _b, _c, _d;
        return trace('wix-data-error-response', {
                responseData: (_a = error === null || error === void 0 ? void 0 : error.response) === null || _a === void 0 ? void 0 : _a.data,
                status: (_b = error === null || error === void 0 ? void 0 : error.response) === null || _b === void 0 ? void 0 : _b.status,
                requestId: (_d = (_c = error === null || error === void 0 ? void 0 : error.response) === null || _c === void 0 ? void 0 : _c.headers) === null || _d === void 0 ? void 0 : _d['x-wix-request-id'],
                error: (error === null || error === void 0 ? void 0 : error.response) ? undefined : error,
            })(function() {
                return Promise.resolve();
            })
            .then(function() {
                var _a;
                if (isPayloadTooLarge(error)) {
                    return payloadTooLargeError();
                }
                var jsonError = (_a = error === null || error === void 0 ? void 0 : error.response) === null || _a === void 0 ? void 0 : _a.data;
                if (isHookError(jsonError)) {
                    return parseHookError(jsonError);
                }
                if (isApplicationError(jsonError)) {
                    return parseApplicationError(jsonError);
                }
                if (isValidationError(jsonError)) {
                    return parseValidationError(jsonError);
                }
                return internalError();
            })
            .then(function(parsedError) {
                return Promise.reject(parsedError);
            });
    };
};

function isPayloadTooLarge(error) {
    var _a;
    return ((_a = error === null || error === void 0 ? void 0 : error.response) === null || _a === void 0 ? void 0 : _a.status) === 413;
}

function payloadTooLargeError() {
    return wixDataError(messages.payloadIsTooLarge(), PayloadTooLargeCode);
}

function isApplicationError(error) {
    var _a;
    return (_a = error === null || error === void 0 ? void 0 : error.details) === null || _a === void 0 ? void 0 : _a.applicationError;
}

function isValidationError(error) {
    var _a;
    return (_a = error === null || error === void 0 ? void 0 : error.details) === null || _a === void 0 ? void 0 : _a.validationError;
}

function isHookError(error) {
    var _a, _b;
    return ((_b = (_a = error === null || error === void 0 ? void 0 : error.details) === null || _a === void 0 ? void 0 : _a.applicationError) === null || _b === void 0 ? void 0 : _b.code) === HooksErrorCode;
}

function internalError() {
    return wixDataError(messages.internalError('Unknown error.'), codes.UnknownError);
}

function parseHookError(error) {
    var data = error.details.applicationError.data;
    if (data === null || data === void 0 ? void 0 : data.jsErrorValue) {
        var _a = data.jsErrorValue,
            name_1 = _a.name,
            message = _a.message,
            code = _a.code;
        var result = new Error(message);
        result.name = name_1;
        // @ts-expect-error
        result.code = code;
        return result;
    }
    return data === null || data === void 0 ? void 0 : data.value;
}
export var parseApplicationError = function(error) {
    var code = mapErrorCode(error.details.applicationError.code);
    return wixDataError(error.details.applicationError.description, code, error.details);
};

function parseValidationError(error) {
    var validationMessages = (error.details.validationError.fieldViolations || []).map(function(v) {
        return "".concat(v.field, ": ").concat(v.description);
    });
    var message = __spreadArray([error.message], validationMessages, true).join('\n');
    return wixDataError(message, ValidationErrorCode, error.details);
}
export var parseBulkError = function(items) {
    return function(error) {
        var _a = getBulkErrorDetails(error),
            message = _a.message,
            code = _a.code,
            name = _a.name;
        return {
            message: message,
            code: code,
            name: name,
            originalIndex: error.originalIndex,
            item: items[error.originalIndex],
        };
    };
};

function getBulkErrorDetails(error) {
    var _a, _b, _c;
    var appError = (_a = error.details) === null || _a === void 0 ? void 0 : _a.applicationError;
    if (isHookError(error)) {
        var jsError = (_b = appError.data) === null || _b === void 0 ? void 0 : _b.jsErrorValue;
        if (jsError) {
            return {
                message: jsError.message,
                code: jsError.code,
                name: jsError.name,
            };
        }
        var value = (_c = appError.data) === null || _c === void 0 ? void 0 : _c.value;
        if (typeof value === 'string') {
            return {
                message: value
            };
        }
        return {};
    }
    return {
        code: mapErrorCode(appError === null || appError === void 0 ? void 0 : appError.code) || codes.UnknownError,
        message: (appError === null || appError === void 0 ? void 0 : appError.description) || error.message,
        name: 'Error',
    };
}
