// @wix/animations-kit/dist/esm/definitions/animation/scroll/growScroll.js
import {
    __rest
} from "tslib";
const name = 'GrowScroll';
const properties = {
    schema: {},
};
const MAX_Y_TRAVEL = 40;
const powerMap = {
    soft: {
        scaleFrom: 0.8,
        scaleTo: 1.2,
        travelY: 0
    },
    medium: {
        scaleFrom: 0.3,
        scaleTo: 1.7,
        travelY: 0.75
    },
    hard: {
        scaleFrom: 0,
        scaleTo: 4,
        travelY: 1
    },
};
const rangeValues = { in: (scaleFrom, _scaleTo, travelY) => ({
        fromValues: {
            scale: scaleFrom,
            travel: travelY
        },
        toValues: {
            scale: 1,
            travel: 0
        },
        scaleDelay: 600,
        scaleEase: 'sineIn',
    }),
    out: (_scaleFrom, scaleTo, travelY) => ({
        fromValues: {
            scale: 1,
            travel: 0
        },
        toValues: {
            scale: scaleTo,
            travel: -travelY
        },
        scaleDelay: 0,
        scaleEase: 'linear',
    }),
    continuous: (scaleFrom, scaleTo, travelY) => ({
        fromValues: {
            scale: scaleFrom,
            travel: travelY
        },
        toValues: {
            scale: scaleTo,
            travel: -travelY
        },
        scaleDelay: 220,
        scaleEase: 'sineInOut',
    }),
};
/**
 * Grow Scroll
 */
function animate(kit, elements, _duration, _delay, _a = {}) {
    var {
        power,
        range = 'in',
        scale = range === 'in' ? powerMap.hard.scaleFrom : powerMap.hard.scaleTo,
        speed = 0
    } = _a, params = __rest(_a, ["power", "range", "scale", "speed"]);
    const sequence = kit.sequence(params);
    const {
        scaleFrom,
        scaleTo,
        travelY
    } = typeof power !== 'undefined' ?
        powerMap[power] :
        {
            scaleFrom: scale,
            scaleTo: scale,
            travelY: speed,
        };
    const {
        fromValues,
        toValues,
        scaleDelay,
        scaleEase
    } = rangeValues[range](scaleFrom, scaleTo, travelY * -MAX_Y_TRAVEL);
    const delay = power === 'hard' && range !== 'out' ? scaleDelay : 0;
    const duration = range === 'out' ? 700 : 1000 - delay;
    sequence.add([
        kit.animate('BaseScale', elements, duration, delay, {
            from: {
                scale: fromValues.scale
            },
            to: {
                scale: toValues.scale
            },
            ease: scaleEase,
        }),
        kit.animate('BasePosition', elements, 1000, 0, {
            from: {
                y: `${fromValues.travel}vh`
            },
            to: {
                y: `${toValues.travel}vh`
            },
            ease: 'linear',
        }),
    ]);
    return sequence.get();
}
const api = {
    getScrubOffsets({
        power,
        range = 'in',
        speed = 0,
    }) {
        const offset = (typeof power !== 'undefined' ?
            powerMap[power].travelY :
            Math.abs(speed)) * MAX_Y_TRAVEL;
        return {
            start: range === 'out' ? '0px' : `${-offset}vh`,
            end: range === 'in' ? '0px' : `${offset}vh`,
        };
    },
};
export {
    name,
    properties,
    animate,
    api
};
