// @wix/cloud-elementory-protocol/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getBackendPackageNameFromImportName = exports.generateBackendPackageWebMethodUrl = exports.generatePackageWebMethodUrl = exports.exceptionToWebMethodPayload = exports.resultToWebMethodPayload = exports.convertFromCustomFormat = exports.convertToCustomFormat = void 0;
var converters_1 = require("./src/converters");
Object.defineProperty(exports, "convertToCustomFormat", {
    enumerable: true,
    get: function() {
        return converters_1.convertToCustomFormat;
    }
});
Object.defineProperty(exports, "convertFromCustomFormat", {
    enumerable: true,
    get: function() {
        return converters_1.convertFromCustomFormat;
    }
});
var web_method_protocol_1 = require("./src/web-method-protocol");
Object.defineProperty(exports, "resultToWebMethodPayload", {
    enumerable: true,
    get: function() {
        return web_method_protocol_1.resultToWebMethodPayload;
    }
});
Object.defineProperty(exports, "exceptionToWebMethodPayload", {
    enumerable: true,
    get: function() {
        return web_method_protocol_1.exceptionToWebMethodPayload;
    }
});
Object.defineProperty(exports, "generatePackageWebMethodUrl", {
    enumerable: true,
    get: function() {
        return web_method_protocol_1.generatePackageWebMethodUrl;
    }
});
Object.defineProperty(exports, "generateBackendPackageWebMethodUrl", {
    enumerable: true,
    get: function() {
        return web_method_protocol_1.generateBackendPackageWebMethodUrl;
    }
});
Object.defineProperty(exports, "getBackendPackageNameFromImportName", {
    enumerable: true,
    get: function() {
        return web_method_protocol_1.getBackendPackageNameFromImportName;
    }
});
